File: kits/data-integration/providers/capture_mode/WebMarkdown.swift
// Data Integration Kit - Web Markdown Capture Provider
// HTML to Markdown conversion with Readability extraction and YAML frontmatter

import Foundation

public final class WebMarkdownCaptureProvider {
    public init() {}

    public func capture(input: CaptureInput, config: CaptureConfig) async throws -> CaptureItem {
        guard let urlString = input.url, let url = URL(string: urlString) else {
            throw CaptureError.missingURL
        }

        let (data, _) = try await URLSession.shared.data(from: url)
        guard let html = String(data: data, encoding: .utf8) else {
            throw CaptureError.parseError("Unable to decode HTML as UTF-8")
        }

        let meta = extractArticleMeta(html: html)
        let cleaned = stripNonContent(html: html)
        let markdown = htmlToMarkdown(html: cleaned)

        let includeFrontmatter = (config.options?["frontmatter"] as? Bool) ?? true
        let frontmatter = includeFrontmatter ? generateFrontmatter(meta: meta, url: urlString) : nil
        let content = frontmatter.map { "\($0)\n\n\(markdown)" } ?? markdown

        var tags = ["markdown"]
        tags.append(contentsOf: meta.tags)

        return CaptureItem(
            content: content,
            sourceMetadata: SourceMetadata(
                title: meta.title,
                url: urlString,
                capturedAt: ISO8601DateFormatter().string(from: Date()),
                contentType: "text/markdown",
                author: meta.author,
                tags: tags,
                source: "web_markdown"
            ),
            rawData: (config.options?["includeHtml"] as? Bool == true) ? html : nil
        )
    }

    public func supports(input: CaptureInput) -> Bool {
        guard let url = input.url else { return false }
        return url.hasPrefix("http://") || url.hasPrefix("https://")
    }

    // MARK: - Article Metadata Extraction

    private struct ArticleMeta {
        var title: String
        var author: String?
        var date: String?
        var description: String?
        var tags: [String]
    }

    private func extractMeta(html: String, property: String) -> String? {
        let escaped = NSRegularExpression.escapedPattern(for: property)
        let pattern = "(?i)<meta[^>]+(?:property|name)=[\"']\(escaped)[\"'][^>]+content=[\"']([^\"']+)[\"']"
        guard let regex = try? NSRegularExpression(pattern: pattern),
              let match = regex.firstMatch(in: html, range: NSRange(html.startIndex..., in: html)),
              let range = Range(match.range(at: 1), in: html) else { return nil }
        return String(html[range]).trimmingCharacters(in: .whitespaces)
    }

    private func extractArticleMeta(html: String) -> ArticleMeta {
        var title = extractMeta(html: html, property: "og:title")
        if title == nil,
           let regex = try? NSRegularExpression(pattern: "(?i)<title>([^<]+)</title>"),
           let match = regex.firstMatch(in: html, range: NSRange(html.startIndex..., in: html)),
           let range = Range(match.range(at: 1), in: html) {
            title = String(html[range]).trimmingCharacters(in: .whitespaces)
        }

        let author = extractMeta(html: html, property: "author")
            ?? extractMeta(html: html, property: "article:author")
        let date = extractMeta(html: html, property: "article:published_time")
            ?? extractMeta(html: html, property: "date")
        let description = extractMeta(html: html, property: "og:description")
            ?? extractMeta(html: html, property: "description")

        let tagsStr = extractMeta(html: html, property: "article:tag")
            ?? extractMeta(html: html, property: "keywords")
        let tags = tagsStr?.components(separatedBy: ",")
            .map { $0.trimmingCharacters(in: .whitespaces) }
            .filter { !$0.isEmpty } ?? []

        return ArticleMeta(title: title ?? "Untitled", author: author,
                          date: date, description: description, tags: tags)
    }

    // MARK: - HTML Cleaning

    private func stripNonContent(html: String) -> String {
        var result = html
        for tag in ["script", "style", "nav", "footer", "header", "aside", "iframe", "noscript"] {
            let pattern = "(?is)<\(tag)[^>]*>[\\s\\S]*?</\(tag)>"
            if let regex = try? NSRegularExpression(pattern: pattern) {
                let range = NSRange(result.startIndex..., in: result)
                result = regex.stringByReplacingMatches(in: result, range: range, withTemplate: "")
            }
        }
        return result
    }

    private func stripTags(_ html: String) -> String {
        html.replacingOccurrences(of: "<[^>]+>", with: "", options: .regularExpression)
    }

    private func decodeEntities(_ text: String) -> String {
        text.replacingOccurrences(of: "&nbsp;", with: " ")
            .replacingOccurrences(of: "&amp;", with: "&")
            .replacingOccurrences(of: "&lt;", with: "<")
            .replacingOccurrences(of: "&gt;", with: ">")
            .replacingOccurrences(of: "&quot;", with: "\"")
            .replacingOccurrences(of: "&#39;", with: "'")
    }

    // MARK: - HTML to Markdown Conversion

    private func htmlToMarkdown(html: String) -> String {
        var md = html

        // Headings h1-h6
        for i in 1...6 {
            let prefix = String(repeating: "#", count: i)
            let pattern = "(?is)<h\(i)[^>]*>([\\s\\S]*?)</h\(i)>"
            if let regex = try? NSRegularExpression(pattern: pattern) {
                let range = NSRange(md.startIndex..., in: md)
                let matches = regex.matches(in: md, range: range).reversed()
                for match in matches {
                    guard let fullRange = Range(match.range, in: md),
                          let contentRange = Range(match.range(at: 1), in: md) else { continue }
                    let content = stripTags(String(md[contentRange])).trimmingCharacters(in: .whitespaces)
                    md.replaceSubrange(fullRange, with: "\n\(prefix) \(content)\n")
                }
            }
        }

        // Bold
        md = replacePattern(md, pattern: "(?is)<(?:strong|b)>([\\s\\S]*?)</(?:strong|b)>") { "**\(self.stripTags($0))**" }
        // Italic
        md = replacePattern(md, pattern: "(?is)<(?:em|i)>([\\s\\S]*?)</(?:em|i)>") { "*\(self.stripTags($0))*" }

        // Links
        if let regex = try? NSRegularExpression(pattern: #"(?is)<a[^>]+href=["']([^"']+)["'][^>]*>([\s\S]*?)</a>"#) {
            let range = NSRange(md.startIndex..., in: md)
            let matches = regex.matches(in: md, range: range).reversed()
            for match in matches {
                guard let fullRange = Range(match.range, in: md),
                      let hrefRange = Range(match.range(at: 1), in: md),
                      let textRange = Range(match.range(at: 2), in: md) else { continue }
                let href = String(md[hrefRange])
                let text = stripTags(String(md[textRange])).trimmingCharacters(in: .whitespaces)
                md.replaceSubrange(fullRange, with: "[\(text)](\(href))")
            }
        }

        // Inline code
        md = replacePattern(md, pattern: "(?is)<code>([\\s\\S]*?)</code>") { "`\(self.decodeEntities($0))`" }

        // Blockquotes
        md = replacePattern(md, pattern: "(?is)<blockquote[^>]*>([\\s\\S]*?)</blockquote>") {
            let lines = self.stripTags($0).trimmingCharacters(in: .whitespacesAndNewlines)
                .components(separatedBy: "\n").map { "> \($0)" }
            return "\n\(lines.joined(separator: "\n"))\n"
        }

        // Unordered lists
        md = replacePattern(md, pattern: "(?is)<ul[^>]*>([\\s\\S]*?)</ul>") {
            self.convertListItems($0, prefix: "- ")
        }

        // Ordered lists
        md = replacePattern(md, pattern: "(?is)<ol[^>]*>([\\s\\S]*?)</ol>") {
            self.convertOrderedListItems($0)
        }

        // Tables
        md = replacePattern(md, pattern: "(?is)<table[^>]*>([\\s\\S]*?)</table>") {
            self.convertTable(html: $0)
        }

        // HR, BR, P
        md = md.replacingOccurrences(of: "<hr\\s*/?>", with: "\n---\n", options: .regularExpression)
        md = md.replacingOccurrences(of: "<br\\s*/?>", with: "\n", options: .regularExpression)
        md = replacePattern(md, pattern: "(?is)<p[^>]*>([\\s\\S]*?)</p>") {
            "\n\(self.stripTags($0).trimmingCharacters(in: .whitespaces))\n"
        }

        md = stripTags(md)
        md = decodeEntities(md)
        md = md.replacingOccurrences(of: "\n{3,}", with: "\n\n", options: .regularExpression)
        return md.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    private func replacePattern(_ string: String, pattern: String, replacement: (String) -> String) -> String {
        guard let regex = try? NSRegularExpression(pattern: pattern) else { return string }
        var result = string
        let matches = regex.matches(in: result, range: NSRange(result.startIndex..., in: result)).reversed()
        for match in matches {
            guard let fullRange = Range(match.range, in: result),
                  let groupRange = Range(match.range(at: 1), in: result) else { continue }
            let content = String(result[groupRange])
            result.replaceSubrange(fullRange, with: replacement(content))
        }
        return result
    }

    private func convertListItems(_ html: String, prefix: String) -> String {
        guard let regex = try? NSRegularExpression(pattern: "(?is)<li[^>]*>([\\s\\S]*?)</li>") else { return "" }
        let range = NSRange(html.startIndex..., in: html)
        let items = regex.matches(in: html, range: range).compactMap { match -> String? in
            guard let r = Range(match.range(at: 1), in: html) else { return nil }
            return "\(prefix)\(stripTags(String(html[r])).trimmingCharacters(in: .whitespaces))"
        }
        return "\n\(items.joined(separator: "\n"))\n"
    }

    private func convertOrderedListItems(_ html: String) -> String {
        guard let regex = try? NSRegularExpression(pattern: "(?is)<li[^>]*>([\\s\\S]*?)</li>") else { return "" }
        let range = NSRange(html.startIndex..., in: html)
        let items = regex.matches(in: html, range: range).enumerated().compactMap { (i, match) -> String? in
            guard let r = Range(match.range(at: 1), in: html) else { return nil }
            return "\(i + 1). \(stripTags(String(html[r])).trimmingCharacters(in: .whitespaces))"
        }
        return "\n\(items.joined(separator: "\n"))\n"
    }

    private func convertTable(html: String) -> String {
        guard let rowRegex = try? NSRegularExpression(pattern: "(?is)<tr[^>]*>([\\s\\S]*?)</tr>"),
              let cellRegex = try? NSRegularExpression(pattern: "(?is)<(?:td|th)[^>]*>([\\s\\S]*?)</(?:td|th)>") else {
            return ""
        }
        let rowRange = NSRange(html.startIndex..., in: html)
        let rowMatches = rowRegex.matches(in: html, range: rowRange)
        var rows: [[String]] = []
        for rowMatch in rowMatches {
            guard let r = Range(rowMatch.range(at: 1), in: html) else { continue }
            let rowContent = String(html[r])
            let cellRange = NSRange(rowContent.startIndex..., in: rowContent)
            let cells = cellRegex.matches(in: rowContent, range: cellRange).compactMap { match -> String? in
                guard let cr = Range(match.range(at: 1), in: rowContent) else { return nil }
                return stripTags(String(rowContent[cr])).trimmingCharacters(in: .whitespaces)
            }
            if !cells.isEmpty { rows.append(cells) }
        }
        guard !rows.isEmpty else { return "" }
        let maxCols = rows.map { $0.count }.max() ?? 0
        var lines: [String] = []
        for (i, row) in rows.enumerated() {
            let padded = (0..<maxCols).map { j in j < row.count ? row[j] : "" }
            lines.append("| \(padded.joined(separator: " | ")) |")
            if i == 0 {
                lines.append("| \((0..<maxCols).map { _ in "---" }.joined(separator: " | ")) |")
            }
        }
        return "\n\(lines.joined(separator: "\n"))\n"
    }

    // MARK: - Frontmatter Generation

    private func generateFrontmatter(meta: ArticleMeta, url: String) -> String {
        var lines = ["---"]
        lines.append("title: \"\(meta.title.replacingOccurrences(of: "\"", with: "\\\""))\"")
        if let author = meta.author { lines.append("author: \"\(author)\"") }
        if let date = meta.date { lines.append("date: \"\(date)\"") }
        lines.append("source: \"\(url)\"")
        if let desc = meta.description {
            lines.append("description: \"\(desc.replacingOccurrences(of: "\"", with: "\\\""))\"")
        }
        if !meta.tags.isEmpty {
            let tagStr = meta.tags.map { "\"\($0)\"" }.joined(separator: ", ")
            lines.append("tags: [\(tagStr)]")
        }
        lines.append("---")
        return lines.joined(separator: "\n")
    }
}

File: kits/data-integration/providers/capture_mode/web_markdown.rs
// Data Integration Kit - Web Markdown Capture Provider
// HTML to Markdown conversion with Readability extraction and YAML frontmatter

use std::collections::HashMap;

pub const PROVIDER_ID: &str = "web_markdown";
pub const PLUGIN_TYPE: &str = "capture_mode";

#[derive(Debug, Clone)]
pub struct CaptureInput {
    pub url: Option<String>,
    pub file: Option<Vec<u8>>,
    pub email: Option<String>,
    pub share_data: Option<serde_json::Value>,
}

#[derive(Debug, Clone)]
pub struct CaptureConfig {
    pub mode: String,
    pub options: Option<HashMap<String, serde_json::Value>>,
}

#[derive(Debug, Clone)]
pub struct SourceMetadata {
    pub title: String,
    pub url: Option<String>,
    pub captured_at: String,
    pub content_type: String,
    pub author: Option<String>,
    pub tags: Option<Vec<String>>,
    pub source: Option<String>,
}

#[derive(Debug, Clone)]
pub struct CaptureItem {
    pub content: String,
    pub source_metadata: SourceMetadata,
    pub raw_data: Option<String>,
}

#[derive(Debug)]
pub enum CaptureError {
    MissingUrl,
    FetchError(String),
    ParseError(String),
}

impl std::fmt::Display for CaptureError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            CaptureError::MissingUrl => write!(f, "web_markdown capture requires a URL"),
            CaptureError::FetchError(e) => write!(f, "Fetch error: {}", e),
            CaptureError::ParseError(e) => write!(f, "Parse error: {}", e),
        }
    }
}

#[derive(Debug, Default)]
struct ArticleMeta {
    title: String,
    author: Option<String>,
    date: Option<String>,
    description: Option<String>,
    tags: Vec<String>,
}

fn extract_meta(html: &str, property: &str) -> Option<String> {
    let pat = format!(
        r#"(?i)<meta[^>]+(?:property|name)=["']{}["'][^>]+content=["']([^"']+)["']"#,
        regex::escape(property)
    );
    regex::Regex::new(&pat).ok()
        .and_then(|re| re.captures(html))
        .and_then(|caps| caps.get(1).map(|m| m.as_str().trim().to_string()))
}

fn extract_article_meta(html: &str) -> ArticleMeta {
    let title = extract_meta(html, "og:title")
        .or_else(|| {
            regex::Regex::new(r"(?i)<title>([^<]+)</title>").ok()
                .and_then(|re| re.captures(html))
                .map(|caps| caps[1].trim().to_string())
        })
        .unwrap_or_else(|| "Untitled".to_string());

    let author = extract_meta(html, "author").or_else(|| extract_meta(html, "article:author"));
    let date = extract_meta(html, "article:published_time").or_else(|| extract_meta(html, "date"));
    let description = extract_meta(html, "og:description").or_else(|| extract_meta(html, "description"));

    let tags_str = extract_meta(html, "article:tag").or_else(|| extract_meta(html, "keywords"));
    let tags = tags_str.map(|t| t.split(',').map(|s| s.trim().to_string()).filter(|s| !s.is_empty()).collect())
        .unwrap_or_default();

    ArticleMeta { title, author, date, description, tags }
}

fn strip_non_content(html: &str) -> String {
    let mut result = html.to_string();
    for tag in &["script", "style", "nav", "footer", "header", "aside", "iframe", "noscript"] {
        let pat = format!(r"(?is)<{0}[^>]*>[\s\S]*?</{0}>", tag);
        if let Ok(re) = regex::Regex::new(&pat) {
            result = re.replace_all(&result, "").to_string();
        }
    }
    result
}

fn strip_tags(html: &str) -> String {
    regex::Regex::new(r"<[^>]+>").unwrap().replace_all(html, "").to_string()
}

fn decode_entities(text: &str) -> String {
    text.replace("&nbsp;", " ").replace("&amp;", "&")
        .replace("&lt;", "<").replace("&gt;", ">")
        .replace("&quot;", "\"").replace("&#39;", "'")
}

fn html_to_markdown(html: &str) -> String {
    let mut md = html.to_string();

    // Headings h1-h6
    for i in 1..=6 {
        let prefix = "#".repeat(i);
        let pat = format!(r"(?is)<h{}[^>]*>([\s\S]*?)</h{}>", i, i);
        if let Ok(re) = regex::Regex::new(&pat) {
            md = re.replace_all(&md, |caps: &regex::Captures| {
                format!("\n{} {}\n", prefix, strip_tags(&caps[1]).trim())
            }).to_string();
        }
    }

    // Bold
    if let Ok(re) = regex::Regex::new(r"(?is)<(?:strong|b)>([\s\S]*?)</(?:strong|b)>") {
        md = re.replace_all(&md, |caps: &regex::Captures| format!("**{}**", strip_tags(&caps[1]))).to_string();
    }

    // Italic
    if let Ok(re) = regex::Regex::new(r"(?is)<(?:em|i)>([\s\S]*?)</(?:em|i)>") {
        md = re.replace_all(&md, |caps: &regex::Captures| format!("*{}*", strip_tags(&caps[1]))).to_string();
    }

    // Links
    if let Ok(re) = regex::Regex::new(r#"(?is)<a[^>]+href=["']([^"']+)["'][^>]*>([\s\S]*?)</a>"#) {
        md = re.replace_all(&md, |caps: &regex::Captures| {
            format!("[{}]({})", strip_tags(&caps[2]).trim(), &caps[1])
        }).to_string();
    }

    // Code blocks
    if let Ok(re) = regex::Regex::new(r#"(?is)<pre[^>]*>\s*<code[^>]*(?:class=["']language-(\w+)["'][^>]*)?>([\\s\\S]*?)</code>\s*</pre>"#) {
        md = re.replace_all(&md, |caps: &regex::Captures| {
            let lang = caps.get(1).map(|m| m.as_str()).unwrap_or("");
            format!("\n```{}\n{}\n```\n", lang, decode_entities(&caps[2]).trim())
        }).to_string();
    }

    // Inline code
    if let Ok(re) = regex::Regex::new(r"(?is)<code>([\s\S]*?)</code>") {
        md = re.replace_all(&md, |caps: &regex::Captures| format!("`{}`", decode_entities(&caps[1]))).to_string();
    }

    // Blockquotes
    if let Ok(re) = regex::Regex::new(r"(?is)<blockquote[^>]*>([\s\S]*?)</blockquote>") {
        md = re.replace_all(&md, |caps: &regex::Captures| {
            let text = strip_tags(&caps[1]).trim().to_string();
            let lines: Vec<String> = text.lines().map(|l| format!("> {}", l)).collect();
            format!("\n{}\n", lines.join("\n"))
        }).to_string();
    }

    // Unordered lists
    if let Ok(ul_re) = regex::Regex::new(r"(?is)<ul[^>]*>([\s\S]*?)</ul>") {
        md = ul_re.replace_all(&md, |caps: &regex::Captures| {
            let items = &caps[1];
            if let Ok(li_re) = regex::Regex::new(r"(?is)<li[^>]*>([\s\S]*?)</li>") {
                let result: Vec<String> = li_re.captures_iter(items)
                    .map(|c| format!("- {}", strip_tags(&c[1]).trim()))
                    .collect();
                return format!("\n{}\n", result.join("\n"));
            }
            String::new()
        }).to_string();
    }

    // Ordered lists
    if let Ok(ol_re) = regex::Regex::new(r"(?is)<ol[^>]*>([\s\S]*?)</ol>") {
        md = ol_re.replace_all(&md, |caps: &regex::Captures| {
            let items = &caps[1];
            if let Ok(li_re) = regex::Regex::new(r"(?is)<li[^>]*>([\s\S]*?)</li>") {
                let result: Vec<String> = li_re.captures_iter(items).enumerate()
                    .map(|(i, c)| format!("{}. {}", i + 1, strip_tags(&c[1]).trim()))
                    .collect();
                return format!("\n{}\n", result.join("\n"));
            }
            String::new()
        }).to_string();
    }

    // Tables
    if let Ok(table_re) = regex::Regex::new(r"(?is)<table[^>]*>([\s\S]*?)</table>") {
        md = table_re.replace_all(&md, |caps: &regex::Captures| convert_table(&caps[1])).to_string();
    }

    // HR
    md = regex::Regex::new(r"(?i)<hr\s*/?>").unwrap().replace_all(&md, "\n---\n").to_string();
    // BR and P
    md = regex::Regex::new(r"(?i)<br\s*/?>").unwrap().replace_all(&md, "\n").to_string();
    if let Ok(re) = regex::Regex::new(r"(?is)<p[^>]*>([\s\S]*?)</p>") {
        md = re.replace_all(&md, |caps: &regex::Captures| format!("\n{}\n", strip_tags(&caps[1]).trim())).to_string();
    }

    md = strip_tags(&md);
    md = decode_entities(&md);
    regex::Regex::new(r"\n{3,}").unwrap().replace_all(&md, "\n\n").trim().to_string()
}

fn convert_table(html: &str) -> String {
    let row_re = regex::Regex::new(r"(?is)<tr[^>]*>([\s\S]*?)</tr>").unwrap();
    let cell_re = regex::Regex::new(r"(?is)<(?:td|th)[^>]*>([\s\S]*?)</(?:td|th)>").unwrap();

    let mut rows: Vec<Vec<String>> = Vec::new();
    for row_caps in row_re.captures_iter(html) {
        let cells: Vec<String> = cell_re.captures_iter(&row_caps[1])
            .map(|c| strip_tags(&c[1]).trim().to_string())
            .collect();
        if !cells.is_empty() { rows.push(cells); }
    }
    if rows.is_empty() { return String::new(); }

    let max_cols = rows.iter().map(|r| r.len()).max().unwrap_or(0);
    let mut lines = Vec::new();
    for (i, row) in rows.iter().enumerate() {
        let padded: Vec<&str> = (0..max_cols).map(|j| row.get(j).map(|s| s.as_str()).unwrap_or("")).collect();
        lines.push(format!("| {} |", padded.join(" | ")));
        if i == 0 {
            lines.push(format!("| {} |", vec!["---"; max_cols].join(" | ")));
        }
    }
    format!("\n{}\n", lines.join("\n"))
}

fn generate_frontmatter(meta: &ArticleMeta, url: &str) -> String {
    let mut lines = vec!["---".to_string()];
    lines.push(format!("title: \"{}\"", meta.title.replace('"', "\\\"")));
    if let Some(ref author) = meta.author { lines.push(format!("author: \"{}\"", author)); }
    if let Some(ref date) = meta.date { lines.push(format!("date: \"{}\"", date)); }
    lines.push(format!("source: \"{}\"", url));
    if let Some(ref desc) = meta.description { lines.push(format!("description: \"{}\"", desc.replace('"', "\\\""))); }
    if !meta.tags.is_empty() {
        let tags_str: Vec<String> = meta.tags.iter().map(|t| format!("\"{}\"", t)).collect();
        lines.push(format!("tags: [{}]", tags_str.join(", ")));
    }
    lines.push("---".to_string());
    lines.join("\n")
}

pub struct WebMarkdownCaptureProvider;

impl WebMarkdownCaptureProvider {
    pub fn new() -> Self { Self }

    pub fn capture(&self, input: &CaptureInput, config: &CaptureConfig) -> Result<CaptureItem, CaptureError> {
        let url = input.url.as_ref().ok_or(CaptureError::MissingUrl)?;
        let html = http_get(url).map_err(|e| CaptureError::FetchError(e.to_string()))?;

        let meta = extract_article_meta(&html);
        let cleaned = strip_non_content(&html);
        let markdown = html_to_markdown(&cleaned);

        let include_frontmatter = config.options.as_ref()
            .and_then(|o| o.get("frontmatter").and_then(|v| v.as_bool()))
            .unwrap_or(true);

        let content = if include_frontmatter {
            format!("{}\n\n{}", generate_frontmatter(&meta, url), markdown)
        } else {
            markdown
        };

        let mut tags = vec!["markdown".to_string()];
        tags.extend(meta.tags.clone());

        Ok(CaptureItem {
            content,
            source_metadata: SourceMetadata {
                title: meta.title,
                url: Some(url.clone()),
                captured_at: chrono::Utc::now().to_rfc3339(),
                content_type: "text/markdown".to_string(),
                author: meta.author,
                tags: Some(tags),
                source: Some("web_markdown".to_string()),
            },
            raw_data: if config.options.as_ref().and_then(|o| o.get("includeHtml")).is_some() {
                Some(html)
            } else {
                None
            },
        })
    }

    pub fn supports(&self, input: &CaptureInput) -> bool {
        input.url.as_ref().map_or(false, |u| {
            u.starts_with("http://") || u.starts_with("https://")
        })
    }
}

fn http_get(url: &str) -> Result<String, CaptureError> {
    Err(CaptureError::FetchError(format!("HTTP client not configured for: {}", url)))
}
